Migrate to GHA (Closes #437)

avalentino opened the pull request:
Set-up CI on GHA and drop Travis-CI.

Tested Python versions: 3.6-3.9 (please let me know if you steel want to test Python 3.5).
Notifications are not active (commented code). You need to set the Gitter token in secrets for that.

Closes #437.

j08lue commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,50 @@
+    - name: Codevov
+      run: |
+        codecov
+#    - name: Notifications
```
I do not have the Gitter secrets - @Fernerkundung or @valgur, do you and could set them up in GH secrets, please?

valgur replied:
I personally don't have any of the keys.

j08lue replied:
I added the `CODECOV_TOKEN` to the secrets.

j08lue replied:
I think I also found the right secrets for Gitter and set them up now. Gitter is not really used, but since we have it, we might as well keep it running for now.

j08lue reviewed (approved):
Nice!

j08lue commented:
We need to change the CI badge in the README, too. 😄

valgur commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,50 @@
+    - name: Test with pytest
+      run: |
+        coverage run --source=sentinelsat -m pytest -v
+    - name: Codevov
```
Could also use https://github.com/codecov/codecov-action here, maybe.
```yaml
      - name: Upload coverage to Codecov
        uses: codecov/codecov-action@v1
        env:
          OS: ${{ matrix.os }}
          PYTHON: ${{ matrix.python-version }}
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          files: ./coverage.xml
          env_vars: OS,PYTHON
          fail_ci_if_error: true
```

avalentino replied:
Done

valgur commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,50 @@
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
```
How about
```yaml
jobs:
  build:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ ubuntu-latest, windows-latest, macos-latest ]
        python-version: ['3.6', '3.7', '3.8', '3.9']
    if: ${{ matrix.os == 'ubuntu-latest' || matrix.python-version == '3.9' }}
```

valgur replied:
Also, the CI tools seem to handle it ok, but it's a good idea to quote the version numbers since they get interpreted as floats otherwise.

avalentino replied:
```
if: ${{ matrix.os == 'ubuntu-latest' || matrix.python-version == '3.9' }}
```

Not sure why but the above seems not to work:

```
Unrecognized named-value: 'matrix'
```

valgur reviewed (changes requested):
Looks good to me in general. Thanks! Left some minor suggestions.

valgur commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,50 @@
+      run: |
+        python -m pip install --upgrade setuptools pip wheel
+        python -m pip install black codecov
+        if [ -f requirements.txt ]; then python -m pip install -r requirements.txt; fi
```
That if statement is pretty pointless and can be removed.

valgur replied:
Actually, I would remove that line altogether, since `requirements.txt` does not get used when installing from PyPI. Would make it a bit more realistic.

valgur commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,50 @@
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade setuptools pip wheel
+        python -m pip install black codecov
```
can be merged with the above command

j08lue commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,53 @@
+        env_vars: OS,PYTHON
+        fail_ci_if_error: true
+#    - name: Notifications
+#      uses: juztcode/gitter-github-action@v1
+#      if: ${{ failure() }}
+#      with:
+#        room-id: ${{ secrets.GITTER_ROOM_ID }}
+#        token: ${{ secrets.GITTER_TOKEN }}
+#        text: sentinelsat build failure
```
Gitter secrets are in place, so you can activate this part.

```suggestion
      uses: juztcode/gitter-github-action@v1
      if: ${{ failure() }}
      with:
        room-id: ${{ secrets.GITTER_ROOM_ID }}
        token: ${{ secrets.GITTER_TOKEN }}
        text: sentinelsat build failure
```

avalentino replied:
This is done now but it seems not to work
```
Error: Request failed with status code 400
```

avalentino commented:
OK, I think I have implemented most of the requested changes.

Using 
```
if: ${{ matrix.os == 'ubuntu-latest' || matrix.python-version == '3.9' }}
```
seems to be not possible, even if it is not clear to me why

I have switched to Python 3.8 for tests on Mac in order to be able to use wheels for fiona.

On windows I have setup a conda environment because fiona wheels seem to be not available on windows
Unfortunately I have 2 errors on windows (only) that are related to XML parsing:

```
xml.etree.ElementTree.ParseError: unclosed token: line 317, column 4
```
I do not have a windows machine so I cannot investigate.
In your experience could it be a problem related to how cassettes have been staved?

j08lue commented:
Great work! The issue with Windows could be due to the way line endings are handled. If that information (in `.gitattributes`) is not correct or the files were committed wrongly, then checking out the code can strip characters off each line in a file (expecting them to be extra carriage returns), which in fact is part of the code... We'll have to dig a bit into that, but I think we could open a separate issue "Tests failing on Windows?" for that.

Regarding Gitter: Could we just remove that part for now, @valgur and @Fernerkundung? We are not active on our Gitter channel at all, so it does not make sense to invest too much effort into this, IMO.

Fernerkundung commented:
+1 on removing the Gitter integration. It is not providing a lot of value right now and can be dropped in my opinion.

avalentino commented:
> Great work! The issue with Windows could be due to the way line endings are handled. If that information (in .gitattributes) is not correct or the files were committed wrongly, then checking out the code can strip characters off each line in a file (expecting them to be extra carriage returns), which in fact is part of the code... We'll have to dig a bit into that, but I think we could open a separate issue "Tests failing on Windows?" for that.

I have not found any `.gitattributes` file.
I have checked the line endings (on my linux box) and I have not noticed anything strange.
Probably this is something that needs to be tested on a windows machine.

If you want I can open a separate issue.
Meanwhile, do you want me to disable tests on windows in this PR?
... to be able to go on and merge.
